### PR TITLE
Support the stable endpoint from MSC2946

### PR DESCRIPTION
**Pull request by clokep**
[MSC2946](https://github.com/matrix-org/matrix-doc/pull/2946) passed FCP and we should now support the stable endpoint to allow clients to use it.

This work is tracked in #10946, but this PR only does the first half of it: adding the stable endpoint and minor bugfixes based on changes to the MSC.

### CODE REVIEW

File: synapse/rest/client/room.py
Comment (DMRobertson) [inline synapse/rest/client/room.py:1165]: I understand "positive" to mean "> 0".

If passing `limit=0` constitutes a valid request, I'd suggest the error message should say "non-negative" (i.e. >=0) instead of "positive".

The terminology is a little awkward, but standard AFAIK.

**Review by DMRobertson — APPROVED** [review]
Seems reasonable! Note for completeness that the corresponding MSC was merged back a fortnight ago.

File: synapse/rest/client/room.py
Comment (DMRobertson) [reply synapse/rest/client/room.py:1165]: Agreed. MSC says

> * **`limit`**: Optional: a client-defined limit to the maximum number of rooms to return per page. Must an integer greater than zero.

So it was correct before.

Comment (clokep) [reply synapse/rest/client/room.py:1165]: I'll back out that commit, must have confused myself at some point!

File: synapse/handlers/room_summary.py
Comment (clokep) [inline synapse/handlers/room_summary.py:280]: I'm not sure if this is the correct ratelimiter to be using or not? It seems most of our other requests use a custom rate limiter instead of the request ratelimiter?

**Review by DMRobertson — APPROVED** [review]
LGTM!

**Comment by clokep** [comment]
I should check if we need to update any tests here.

File: synapse/handlers/room_summary.py
Comment (DMRobertson) [reply synapse/handlers/room_summary.py:280]: Not sure here---might be worth another pair of eyes?

**Comment by clokep** [comment]
I'm converting this to a draft -- when setting the complement tests to use the new endpoint they were failing. I want to investigate before merging this.

File: synapse/app/homeserver.py
Comment (clokep) [inline synapse/app/homeserver.py:197]: For the generic worker it seems we just put `client_resource` under `/_matrix/client` once. I'm not sure why we don't do that here too?

File: synapse/federation/federation_client.py
Comment (clokep) [inline synapse/federation/federation_client.py:1419]: This is the same pattern we do with e.g. `send_join`.

File: synapse/federation/transport/client.py
Comment (clokep) [inline synapse/federation/transport/client.py:1196]: We could choose whether to v1 vs. unstable based on a parameter but the other bits in this file seem to have separate methods for v1 vs. v2, etc. so I followed that pattern. I think we can clean that up separately if it makes sense?

**Comment by clokep** [comment]
See updated complement tests: matrix-org/complement#225.

I've also prepped a branch that finishes off #10946 and double checked that tests still pass.

File: .github/workflows/tests.yml
Comment (DMRobertson) [inline .github/workflows/tests.yml:377]: There's duplication here with `complement.sh`. (The latter of which now has a merge conflict, I think).

Is there some way to make both of these places read the excluded tags from a single file?

Possibly out-of-scope for this change.

File: docs/workers.md
Comment (DMRobertson) [inline docs/workers.md:213]: Does this want to also say "v3" now that Matrix 1.1 is specced?

File: docs/workers.md
Comment (DMRobertson) [inline docs/workers.md:226]: ditto

File: synapse/rest/client/room.py
Comment (DMRobertson) [inline synapse/rest/client/room.py:1148]: do we want `v3` here?

**Review by DMRobertson — COMMENTED** [review]
Also seems reasonable after the rebase! I would like input from @matrix-org/synapse-core re https://github.com/matrix-org/synapse/pull/11329/files#r749355348 though.

**Review by DMRobertson — CHANGES_REQUESTED** [review]
At the very least, that merge conflict needs resolving though!

File: .github/workflows/tests.yml
Comment (clokep) [reply .github/workflows/tests.yml:377]: It is really annoying, see matrix-org/complement#185. I don't have good ideas how to solve it, but I think it is out of scope of this PR.

File: docs/workers.md
Comment (clokep) [reply docs/workers.md:213]: No, new endpoints star at `v1` (note that this matches what is in the MSC also).

Comment (clokep) [reply docs/workers.md:213]: Also federation endpoints have always been versioned per-endpoint so have always started at `v1` (they never had `r0` paths).

File: docs/workers.md
Comment (clokep) [reply docs/workers.md:226]: No, new endpoints start at `v1` (note that this matches what is in the MSC also).

File: synapse/rest/client/room.py
Comment (clokep) [reply synapse/rest/client/room.py:1148]: No, see above.

File: .github/workflows/tests.yml
Comment (DMRobertson) [reply .github/workflows/tests.yml:377]: Glad to see we're both annoyed by this! Agreed, let's leave it.

File: synapse/app/homeserver.py
Comment (richvdh) [reply synapse/app/homeserver.py:197]: I think it's because of `WEB_CLIENT_PREFIX`, down at line 274.

Comment (richvdh) [reply synapse/app/homeserver.py:197]: (which is not to say there aren't way better ways of doing this and I would love it if this stuff were less of a mess. see also https://github.com/matrix-org/synapse/issues/5118)

File: synapse/handlers/room_summary.py
Comment (richvdh) [reply synapse/handlers/room_summary.py:280]: It's not really appropriate, because `get_request_ratelimiter` is configured via the `rc_message` config settings, which are about the number of messages that can be sent.

Let's not introduce another set of config settings just for this, though. Either hardcode it for now or introduce some sort of "generic request" ratelimit setting.

File: synapse/rest/client/room.py
Comment (richvdh) [inline synapse/rest/client/room.py:1150]: ugh, I hate client_patterns. this is shorter and clearer to write out longhand.

Comment (clokep) [reply synapse/rest/client/room.py:1150]: I'd prefer to tackle that separately if we wish to change it instead of tacking it onto an unrelated PR.

Comment (richvdh) [reply synapse/rest/client/room.py:1150]: it's not an unrelated PR - you're adding a new use of `client_patterns` when it would be shorter and clearer just to write it out.

But I'll not block this PR on it if you disagree.

Comment (clokep) [reply synapse/rest/client/room.py:1150]: Do we have it documented that we should not be using `client_patterns` for new code? This is the first I've heard of this.

Comment (richvdh) [reply synapse/rest/client/room.py:1150]: > Do we have it documented that we should not be using client_patterns for new code?

That's not *quite* what I said - there may be times it's useful. What I *am* saying is that this:

```python
    PATTERNS = (
        re.compile(
            "^/_matrix/client/unstable/org.matrix.msc2946"
            "/rooms/(?P<room_id>[^/]*)/hierarchy$"
        ),
        re.compile(
            "^/_matrix/client/v1/rooms/(?P<room_id>[^/]*)/hierarchy$"
        ),
    )
```

is clearer and shorter than:

```python
    PATTERNS = [
        re.compile(
            "^/_matrix/client/unstable/org.matrix.msc2946"
            "/rooms/(?P<room_id>[^/]*)/hierarchy$"
        ),
    ] + list(
        client_patterns(
            "/rooms/(?P<room_id>[^/]*)/hierarchy$", releases=("v1",), unstable=False
        )
    )
```

... ie, I don't really feel like `client_patterns` is giving any value here. Just because it exists doesn't mean we should shoehorn it in everywhere we can.

Still, I'll not object if someone wants to go and rid the codebase of it :)

Comment (richvdh) [reply synapse/rest/client/room.py:1150]: > you can easily use client_patterns, which is the standard way we describe these URLs.

well, sometimes.

>  I would find it confusing to switch to client_patterns when removing the unstable API.

I would certainly not recommend doing that.

File: synapse/handlers/room_summary.py
Comment (clokep) [reply synapse/handlers/room_summary.py]: I went with the hard-coded settings for now (that are fairly high).

**Comment by clokep** [comment]
@DMRobertson Can you take another look over this?

File: synapse/rest/client/room.py
Comment (DMRobertson) [inline synapse/rest/client/room.py:1143]: Paranoid: do we refer to capture groups by their index anywhere? The `P<room_id>` makes it look like we always use named capture groups so probably not.

As a sanity check, I tried grepping for `\|unstable`, but didn't find any matches outside of `configure_workers_and_start`. I guess this is the CLIENT_PATTERNS stuff discussed here and elsewhere. (I don't have an opinion on that and am happy to go with the consensus.)

File: synapse/handlers/room_summary.py
Comment (DMRobertson) [inline synapse/handlers/room_summary.py:98]: Where did these numbers come from, out of interest?

Comment (clokep) [reply synapse/handlers/room_summary.py:98]: I made them up, they should allow for 5 queries per second without enforcing any ratelimiting until 10 queries. This might be a bit high, but I'm never quite sure what to put here and usually err on the side of allowing more. 🤷

File: synapse/rest/client/room.py
Comment (clokep) [reply synapse/rest/client/room.py:1143]: Good question! I wasn't entirely sure, but we end up only passing on [`groupdict()`](https://docs.python.org/3/library/re.html#re.Match.groupdict) from the resulting `Match` object, which does only pass *named* subgroups.

See https://github.com/matrix-org/synapse/blob/b64b6d12d47c012a9996a2585a969d3b962d25fa/synapse/http/server.py#L419-L425

Comment (clokep) [reply synapse/rest/client/room.py:1143]: (I can change it to the suggestion of two different URL paths if you'd prefer, which seems to be what we do in other places. I think there's a benefit to having it match the docs too though!)

Comment (DMRobertson) [reply synapse/rest/client/room.py:1143]: > (I can change it to the suggestion of two different URL paths if you'd prefer, which seems to be what we do in other places. I think there's a benefit to having it match the docs too though!)

I think it's fine as it is, and matching docs is a nice-to-have!

File: synapse/handlers/room_summary.py
Comment (DMRobertson) [reply synapse/handlers/room_summary.py:98]: Fair enough. We can always tighten the limits!

(Though there's a part of me that wonders if rate limiting ought to be handled by the load balancer to keep Synapse itself simpler.)

Comment (clokep) [reply synapse/handlers/room_summary.py:98]: Some rate limits can be tweaked by the Synapse config, FWIW. They can also depend on particular data, e.g. you could limit the message rate per user, per room, per message type (or something like that).

File: synapse/handlers/room_summary.py
Comment (DMRobertson) [reply synapse/handlers/room_summary.py]: >  It's not really appropriate, because get_request_ratelimiter is configured via the rc_message config settings, which are about the number of messages that can be sent.

Maybe a rename would help express this? (Messages as in to_device messages or events?) Not one for here though.

**Review by DMRobertson — APPROVED** [review]
LGTM!

File: docs/workers.md
Comment (daenney) [inline docs/workers.md:212]: Shouldn't this one get `v1` too in that case?

Comment (clokep) [reply docs/workers.md:212]: No, this endpoint was not stabilized and will be removed in the future.

Comment (daenney) [reply docs/workers.md:212]: Cool, thanks!

Comment (clokep) [reply docs/workers.md:212]: No problem! It was still being hit fairly actively on matrix.org last time I checked unfortunately. It was essentially replaced with the `/hierarchy` version.

#10946 tracks this work for cross-referencing.

**Pull request merged**
